controller/internal/controller/servicedeployment_controller.go

```diff
@@ -48,6 +50,11 @@ type ServiceReconciler struct {
 
 func (r *ServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	logger := log.FromContext(ctx)
+	if err := r.ConsoleClient.UseNamespaceCredentials(req.Namespace, r.CredentialsCache); err != nil {
```

Comment:
This is going to have severe race condition issues since the console client is a singleton throughout the entirety of this process.  if there are multiple reconcilers (or even one reconciler run in multiple goroutines) executing simultaneously, they could end up using the wrong namespace's credentials.

I think the safest way is to refactor that to return a `client, err` tuple and then use the client in all following code in the individual reconciler from there (the `ServiceReconciler` in this case).  That way the instance is goroutine-local and no race conditions are a concern.

Reply:
I'm not sure it's actually possible, but if you could keep the console client on the reconcilers passed `ctx` that also might be a way to store the global state, or could think of some other mechanic too, it's unfortunate since a ton of code will need to be refactored with a client object as a return value.